modules/core/src/main/scala/lucuma/odb/api/model/TargetModel.scala

```diff
@@ -58,56 +44,103 @@ object TargetModel extends TargetOptics {
 
   }
 
-  private def createTarget[F[_]: Monad, T](
-    db:         DatabaseState[T],
-    targetId:   Option[Target.Id],
-    programIds: Option[List[Program.Id]],
-    gemTarget:  ValidatedInput[Target]
-  )(
-    implicit S: Stateful[F, T]
-  ): F[ValidatedInput[TargetModel]] =
-     for {
-       i  <- db.target.getUnusedId[F](targetId)
-       ps <- programIds.toList.flatten.traverse(db.program.lookupValidated[F])
-       t   = (i, ps.sequence, gemTarget).mapN { (iʹ, _, tʹ) =>
-         TargetModel(iʹ, Existence.Present, tʹ)
-       }
-       _  <- db.target.saveIfValid[F](t)(_.id)
-     } yield t
+  /**
+   * Input required to create either a non-sidereal or sidereal target.
+   */
+  final case class Create(
+    nonsidereal: Option[CreateNonsidereal],
+    sidereal:    Option[CreateSidereal]
+  ) {
+
+    def name: Option[NonEmptyString] =
+      nonsidereal.map(_.name).orElse(sidereal.map(_.name))
+
+    val toGemTarget: ValidatedInput[Target] =
+      ValidatedInput.requireOne(
+        "create",
+        nonsidereal.map(_.toGemTarget),
+        sidereal.map(_.toGemTarget)
+      )
+
+    def editTargetMap(
+      targetMap:     TargetMap,
+      listName:      String,
+      observationId: Observation.Id
+    ): ValidatedInput[TargetMap] =
+      ValidatedInput.requireOne(
+        "create",
+        nonsidereal.map(_.editTargetMap(targetMap, listName, observationId)),
+        sidereal.map(_.editTargetMap(targetMap, listName, observationId))
+      )
+
+  }
+
+  object Create {
+
+    implicit val DecoderCreate: Decoder[Create] =
+      deriveDecoder[Create]
+
+    implicit val EqCreate: Eq[Create] =
+      Eq.by { a => (
+        a.nonsidereal,
+        a.sidereal
+      )}
+
+    def nonsidereal(n: CreateNonsidereal): Create =
+      Create(n.some, None)
+
+    def sidereal(s: CreateSidereal): Create =
+      Create(None, s.some)
+
+  }
+
+  sealed trait TargetCreator {
+    def name: NonEmptyString
+
+    def toGemTarget: ValidatedInput[Target]
+
+    def editTargetMap(
+      targetMap:     TargetMap,
+      listName:      String,
+      observationId: Observation.Id
+    ): ValidatedInput[TargetMap] = {
+
+      def alreadyExists: InputError =
+        InputError.fromMessage(
+          s"Cannot create a new ${listName.capitalize} target with name '$name' because one already exists in observation ${Gid[Observation.Id].show(observationId)}"
+        )
+
+      (toGemTarget, targetMap.get(name).as(alreadyExists).toInvalidNec(targetMap)).mapN { (t, m) =>
```

Comment:
I guess there should be a way  to delete a target right?

Reply:
That's in the `EditTargetAction` with its `add`, `delete`, and `edit` options.  Maybe it needs a better name now that I think about it.